jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -182,41 +182,39 @@ public static URL computePURootURL(URL pxmlURL, String descriptorLocation) throw
             // e.g. file:/tmp/META-INF/persistence.xml
             // 210280: any file url will be assumed to always reference a file (not a directory)
             result = new URL(pxmlURL, path.toString()); // NOI18N
-        } else if("jar".equals(protocol)) { // NOI18N
-            // e.g. jar:file:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            JarURLConnection conn =
-                    JarURLConnection.class.cast(pxmlURL.openConnection());
-            result = conn.getJarFileURL();
-        } else if("zip".equals(protocol)) { // NOI18N
-            // e.g. zip:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            // stolen from java.net.JarURLConnection.parseSpecs method
-            String spec = pxmlURL.getFile();
-            int separator = spec.lastIndexOf("!/");
-            if (separator == -1) {
-                separator = spec.length() - 1;
-            }
-            result = new File(spec.substring(0, separator++)).toURL();
-        } else if("wsjar".equals(protocol)) { // NOI18N
-            // e.g. wsjar:file:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            // but WS gives use jar:file:..., so we need to match it.
-            String spec = pxmlURL.getFile();
+        } else if("zip".equals(protocol) ||
+                  "jar".equals(protocol) ||
+                  "wsjar".equals(protocol)) {
```

Comment:
Id rather you not mess with ```wsjar``` please. I referenced that implementation because it seemed to already account for the usecases we want covered by ```jar```, such as "path/to/a.war!/WEB-INF/classes/ as per JPA 2.1 Spec section 8.2". But unless you added a new, valid test for ```wsjar``` and it is failing, I'd rather you just leave that implementation alone. Call me paranoid, but we haven't had any issues that I know of so Id rather keep it that way :/